Cleanup Resource IO

lorban opened the pull request:
Refactorings to rationalize and simplify how we do IO with `Resource`s internally.

Ideally, a new `Resource.asContentSource()` method should have been introduced but this is not possible because of packaging: `Resource` is in `jetty.util` while `Content.Source` is in `jetty.io`. So I introduced the `IOResources` helper instead in which all `Resource`-related IO now lives: reading into a `RetainableByteBuffer`, wrapping as a `Content.Source` and copying to a `Content.Sink`. `IOResources` is smart enough to use optimized implementations depending on the resource type and only rely on `Resource.newInputStream()` as a fallback.

See #10741

lorban committed:
cleanup Resource IO

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

joakime commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/URLResourceFactory.java:
```diff
@@ -103,7 +103,7 @@ public Resource newResource(final URI uri)
     }
 
-    private static class URLResource extends Resource
+    public static class URLResource extends Resource
```
EEK!
Please keep this private.
We don't want people using this class directly.

lorban replied:
That was accidental; reverted.

joakime commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/IOResources.java:
```diff
@@ -0,0 +1,139 @@
+import org.eclipse.jetty.util.IteratingCallback;
+import org.eclipse.jetty.util.resource.Resource;
+
+public class IOResources
```
Cool. :+1:

gregw replied:
javadoc

gregw commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/Resource.java:
```diff
@@ -235,20 +234,6 @@ public InputStream newInputStream() throws IOException
-     * @return a readable {@link java.nio.channels.ByteChannel} to the resource or null if one is not available.
-     * @throws IOException if unable to open the readable bytechannel for the resource.
-     */
-    public ReadableByteChannel newReadableByteChannel() throws IOException
```
We can only deprecate and mark for removal in a dot release

lorban replied:
That was accidentally removed; I've restored the method but it is now deprecated and always returns null.

gregw commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/ResourceService.java:
```diff
@@ -717,9 +713,16 @@ protected void writeHttpContent(Request request, Response response, Callback cal
             else
-                new ContentWriterIteratingCallback(content, response, callback).iterate();
+            {
+                ByteBufferPool bufferPool = request.getComponents().getByteBufferPool();
+                int outputBufferSize = request.getConnectionMetaData().getHttpConfiguration().getOutputBufferSize();
+                boolean useOutputDirectByteBuffers = request.getConnectionMetaData().getHttpConfiguration().isUseOutputDirectByteBuffers();
+                IOResources.copy(content.getResource(), bufferPool, outputBufferSize, useOutputDirectByteBuffers, response, callback);
+            }
```
I think the following is actually more readable:
```suggestion
                IOResources.copy(
                    content.getResource(), 
                    request.getComponents().getByteBufferPool(), 
                    request.getConnectionMetaData().getHttpConfiguration().getOutputBufferSize(), 
                 
   request.getConnectionMetaData().getHttpConfiguration().isUseOutputDirectByteBuffers(),
                   response,
                   callback);
            }
```

lorban replied:
Done.

gregw commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/BufferUtil.java:
```diff
@@ -1099,19 +1099,24 @@ public static ByteBuffer toBuffer(Resource resource, boolean direct) throws IOEx
+            int pos = BufferUtil.flipToFill(buffer);
+            buffer.put(b);
+            BufferUtil.flipToFlush(buffer, pos);
+        }
```
It feels like this should be a copy operation

lorban replied:
Only tests use this method. Let me see if I can move it elsewhere or improve it.

gregw reviewed (commented):
Just some comments at this stage:

lorban committed:
more cleanups

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
enhance copy + add todo

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
add javadoc

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
fix javadoc

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into jetty-12.0.x-10741-Resource-as-ContentSource

lorban committed:
fix copy of EOF

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
fix flaky test

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
fix checkstyle

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into jetty-12.0.x-10741-Resource-as-ContentSource

lorban committed:
remove ATOMIC_MOVE flag to fix tests

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban commented:
@gregw @joakime I think I reached a point where this implementation starts making sense despite probably still having rough edges.

I'd be glad if you could concentrate on commenting on the design as much as the implementation.

Thanks!

lorban commented:
@sbordet @janbartel your feedback on the design would be welcome too as this change has a wide impact. Thanks!

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/MultiPartByteRanges.java:
```diff
@@ -135,78 +130,6 @@ public boolean addPart(MultiPart.Part part)
-            return read == byteRange.getLength();
-        }
-    }
-
```
These were public classes, I don't think they can be removed.

lorban replied:
Indeed, I restored and deprecated them.

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/content/CachingHttpContentFactory.java:
```diff
@@ -329,43 +328,55 @@ public CachedHttpContent(String key, HttpContent httpContent)
                     {
-                        try (ReadableByteChannel readableByteChannel = httpContent.getResource().newReadableByteChannel())
+                        Content.Source contentSource = IOResources.asContentSource(httpContent.getResource(), _bufferPool, (int)_contentLengthValue, _useDirectByteBuffers);
+                        chunk = contentSource.read();
```
Is this code assuming that a single read into a `Chunk` returns all the bytes?
I would not make such assumption.

lorban replied:
It's a direct translation of the `ReadableByteChannel` equivalent, which I agree is wrong.
I think `IOResources` could offer an API which uses an aggregator to read a resource into a RetainableByteBuffer.
Let me do that.

lorban replied:
Done.

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/IOResources.java:
```diff
@@ -0,0 +1,343 @@
+     */
+    public static void copy(Content.Source source, Content.Sink sink, Predicate<Throwable> onTransientError, Callback callback)
+    {
+        new ContentCopierIteratingCallback(source, sink, onTransientError, callback).iterate();
```
Can you not use `Content.copy()` instead?

lorban replied:
I should have; it's now done.

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/IOResources.java:
```diff
@@ -0,0 +1,343 @@
+     * <p>A specialized {@link PathContentSource}
+     * whose content is sliced by a byte range.</p>
+     */
+    private static class RangedPathContentSource extends PathContentSource
```
There was identical functionality in `MultiPartByteRanges`, but with bits left out such as `ByteRange`.
Why are we moving this to `oej.io`, as byte ranges seems to be just an HTTP concept?

lorban replied:
True, `ByteRange` is an HTTP concept but it has to be implemented at the IO layer, and this PR is about centralizing IO concepts related to Resource so I think it's fair that the implementation moved to `IOResources`.

lorban committed:
rework IOResources API and usage

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into jetty-12.0.x-10741-Resource-as-ContentSource

janbartel commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/IOResources.java:
```diff
@@ -0,0 +1,343 @@
+     * @param resource the resource from which to get an {@link InputStream}.
+     * @return the {@link InputStream}.
+     */
+    public static InputStream asInputStream(Resource resource)
```
I'm not in favour of mandating that all calls to `Resource.newInputStream()` are replaced by calls to this method instead. Feels like `Resource.newInputStream` should be fixed instead.

lorban replied:
That's a good point, but I wish I could remove `Resource.newInputStream()` entirely.

gregw commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/IOResources.java:
```diff
@@ -0,0 +1,343 @@
+     * @param direct the directness of the buffers, this parameter is ignored if {@code bufferSize} is &lt; 1.
+     * @return the {@link Content.Source}.
+     */
+    public static Content.Source asContentSource(Resource resource, ByteBufferPool bufferPool, int bufferSize, boolean direct)
```
This feels like an anti-pattern in an object oriented world.   Just implement `Resource.asContentSource` already and then each class can do it's best to give an optimal source and no need for  all the instanceof checks and down casting.

lorban replied:
That would have been best but unfortunately, `Resource` is in `jetty-util` while `Content.Source` is in `jetty-io`, so I cannot add a `asContentSource()` method to `Resource`.

gregw commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/IOResources.java:
```diff
@@ -0,0 +1,343 @@
+     * @param length the length of the content to read.
+     * @return the {@link Content.Source}.
+     */
+    public static Content.Source asContentSource(Resource resource, ByteBufferPool bufferPool, int bufferSize, boolean direct, long first, long length)
```
ditto

gregw commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/IOResources.java:
```diff
@@ -0,0 +1,343 @@
+            return contentSource;
+        }
+
+        // TODO MemoryResource could be optimized too
```
Can you do this now, else it will be forgotten.

lorban replied:
Done.

gregw commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/IOResources.java:
```diff
@@ -0,0 +1,343 @@
+        // Fallback to wrapping InputStream.
+        try
+        {
+            return new InputStreamContentSource(resource.newInputStream());
```
newInputStream can be null, so we should probably check for that and fail nicely

Also need to javadoc how this fails. Can it return null? or will it always throw.

lorban replied:
I think throwing is the right thing to do here. I've added that check and updated javadoc.

lorban committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into jetty-12.0.x-10741-Resource-as-ContentSource

lorban committed:
add IOResources.toRetainableByteBuffer as per review comments + add opimizations and extra checks

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
update javadoc

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
Merge branch 'jetty-12.0.x-10741-Resource-as-ContentSource' of github.com:jetty/jetty.project into jetty-12.0.x-10741-Resource-as-ContentSource

lorban committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into jetty-12.0.x-10741-Resource-as-ContentSource

lorban committed:
restore API classes after deprecating them

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
fix checkstyle

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
fix test

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
fix toRetainableByteBuffer

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
fix copy and add tests

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/MultiPartByteRanges.java:
```diff
@@ -138,7 +139,9 @@ public boolean addPart(MultiPart.Part part)
      * whose content is sliced by a byte range.</p>
+     * @deprecated Use {@link IOResources#asContentSource(Resource, ByteBufferPool, int, boolean, long, long)} instead.
      */
+    @Deprecated(since = "12.0.8", forRemoval = true)
```
Please remove deprecation, this class is useful on its own, and users are allowed to subclass it and override methods.

lorban replied:
Agreed & done.

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/MultiPartByteRanges.java:
```diff
@@ -166,7 +169,9 @@ protected int fillBufferFromInputStream(InputStream inputStream, byte[] buffer)
      * whose content is sliced by a byte range.</p>
+     * @deprecated Use {@link IOResources#asContentSource(Resource, ByteBufferPool, int, boolean, long, long)} instead.
      */
+    @Deprecated(since = "12.0.8", forRemoval = true)
```
Ditto.

lorban replied:
Agreed & done.

sbordet commented on jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/HttpOutput.java:
```diff
@@ -1176,7 +1178,9 @@ public void sendContent(InputStream in) throws IOException
      * @throws IOException if the send fails
+     * @deprecated use {@link #sendContent(Resource)} instead.
      */
+    @Deprecated(since = "12.0.7", forRemoval = true)
```
Since 12.0.8.
Please reconsider deprecation too? Users may have got a `ByteChannel` from somewhere, cast to `HttpOutput` and use this method to efficiently copy between the 2.

Please also check the version of this method that takes a callback, either both deprecated or both not so.

lorban replied:
This shouldn't have been deprecated. We should not use this API internally but it is nevertheless still useful for end users so I've restored it.

lorban committed:
remove @deprecate

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
- remove all BufferUtil references to Resource
- undeprecate HttpOutput.sendContent(ByteChannel) and restore tests
- optimize IOResources.copy() backed by path so it is capable of doing single last write when the buffer is large enough

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
fix and cleanup sendContent of ByteBuffer, Resource and HttpContent

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into jetty-12.0.x-10741-Resource-as-ContentSource

janbartel commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/IOResources.java:
```diff
@@ -0,0 +1,525 @@
+
+    /**
+     * <p>Gets an {@link InputStream} with the contents of a resource.</p>
+     * <p>The resource must not be a directory, must exists and there must return non-null to {@link Resource#newInputStream()}.</p>
```
```suggestion
     * <p>The resource must not be a directory, must exist and must return non-null to {@link Resource#newInputStream()}.</p>
```

lorban replied:
Done.

janbartel commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/IOResources.java:
```diff
@@ -0,0 +1,525 @@
+
+    /**
+     * <p>Performs an asynchronous copy of the contents of a resource to a sink, using the given buffer pool and buffer characteristics.</p>
+     * <p>The resource must not be a directory, must exists and there must be
```
```suggestion
     * <p>The resource must not be a directory, must exist and there must be
```

lorban replied:
Done.

janbartel commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/IOResources.java:
```diff
@@ -0,0 +1,525 @@
+
+    /**
+     * <p>Performs an asynchronous copy of a subset of the contents of a resource to a sink, using the given buffer pool and buffer characteristics.</p>
+     * <p>The resource must not be a directory, must exists and there must be
```
```suggestion
     * <p>The resource must not be a directory, must exist and there must be
```

lorban replied:
Done.

janbartel commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/IOResources.java:
```diff
@@ -0,0 +1,525 @@
+     * <p>The resource must not be a directory, must exists and there must be
+     * a way to access its contents.</p>
+     * <p>Multiple optimized methods are used to access the resource's contents but if they all fail,
+     * {@link Resource#newInputStream()} is used as a fallback.</p>
```
Isn't `ContentSource` the fallback? I don't see any calls to `Resource.newInputStream`.

lorban replied:
`Resource.newInputStream` is the fallback of the fallback, but you're right: `asContentSource()` is what should be documented as the fallback.

janbartel commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/IOResources.java:
```diff
@@ -0,0 +1,525 @@
+     * <p>The resource must not be a directory, must exists and there must be
+     * a way to access its contents.</p>
+     * <p>Multiple optimized methods are used to access the resource's contents but if they all fail,
+     * {@link Resource#newInputStream()} is used as a fallback.</p>
```
Isn't `ContentSource` the fallback? I don't see any call to `Resource.newInputStream`?

lorban replied:
`Resource.newInputStream` is the fallback of the fallback, but you're right: `asContentSource()` is what should be documented as the fallback.

lorban committed:
fix doc

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

janbartel commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/CombinedResource.java:
```diff
@@ -343,6 +343,7 @@ public void copyTo(Path destination) throws IOException
                     // use old school stream based copy
                     try (InputStream in = r.newInputStream(); OutputStream out = Files.newOutputStream(pathTo))
                     {
+                        // TODO this is very inefficient as copy() allocates a 64K buffer.
```
Not changing the above line to use `IOResources.newInputStream`?

lorban replied:
This is in `java-util`, so here it's not possible to use `IOResources` which is in `java-io`.

janbartel commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/PathResource.java:
```diff
@@ -347,8 +347,7 @@ public long length()
         }
         catch (IOException e)
         {
-            // in case of error, use Files.size() logic of 0L
```
The comment doesn't match the value.

lorban replied:
That's why I deleted the comment; I changed this implementation because all the other ones were returning -1, and I saw no reason why `PathResource` should differ.

sbordet commented on documentation/jetty-documentation/src/main/java/org/eclipse/jetty/docs/programming/server/http2/HTTP2ServerDocs.java:
```diff
@@ -301,7 +302,7 @@ public void push() throws Exception
         // tag::push[]
         // The favicon bytes.
-        ByteBuffer faviconBuffer = BufferUtil.toBuffer(ResourceFactory.root().newResource("/path/to/favicon.ico"), true);
+        ByteBuffer faviconBuffer = BufferUtil.toMappedBuffer(Paths.get("/path/to/favicon.ico"));
```
Do not want to encourage use of mapped buffers in the docs.
Keep it simple and back to normal buffers.

lorban replied:
Reverted.

lorban committed:
handle review comments: improve javadoc

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/IOResources.java:
```diff
@@ -398,11 +398,17 @@ protected Action process() throws Throwable
-            if (remainingLength >= 0)
-                remainingLength -= byteBuffer.remaining();
-            terminated = read == -1 || remainingLength == 0;
+            terminated = eof || remainingLength == 0;
```
Below in `onCompleteSuccess()` you can use `IO.close()` instead of the try/catch.
Probably the same for `onCompleteFailure()`, unless we really want to chain the exception (but it's just a close exception while handling another exception -- kinda who cares).

lorban replied:
Done.

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/BufferUtil.java:
```diff
@@ -1091,35 +1090,6 @@ public static ByteBuffer toBuffer(byte[] array, int offset, int length)
         return ByteBuffer.wrap(array, offset, length);
     }
 
-    public static ByteBuffer toBuffer(Resource resource, boolean direct) throws IOException
```
Can't remove `public` methods.

lorban replied:
You're right; reverted.

sbordet commented on jetty-core/jetty-util/src/test/java/org/eclipse/jetty/util/BufferUtilTest.java:
```diff
@@ -349,26 +344,4 @@ private void testWriteToWithBufferThatDoesNotExposeArray(int capacity) throws IO
     }
-
-    @Test
-    public void testToMappedBufferResource() throws Exception
```
Restore if the methods in `BufferUtil` are restored.

lorban replied:
You're right; reverted.

sbordet commented on jetty-core/jetty-util/src/test/java/org/eclipse/jetty/util/resource/UrlResourceFactoryTest.java:
```diff
@@ -308,4 +307,14 @@ private static long fileSize(URL url) throws IOException
+            byte[] buffer = new byte[Math.toIntExact(resource.length())];
+            is.read(buffer);
+            return ByteBuffer.wrap(buffer);
+        }
```
This is only reading part of the `InputStream`.

lorban replied:
I changed that code to using `BufferUtil` instead.

lorban committed:
restore API

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into jetty-12.0.x-10741-Resource-as-ContentSource

lorban committed:
handle review comments

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
fix tests + update depreciation

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into jetty-12.0.x-10741-Resource-as-ContentSource

gregw commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/MultiPartByteRanges.java:
```diff
@@ -231,19 +231,8 @@ public Part(HttpFields headers, Resource resource, ByteRange byteRange)
-            {
-                throw new RuntimeIOException(e);
-            }
+            // TODO use a buffer pool
```
Can you do this in this PR?

lorban replied:
OK.

gregw commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/IOResources.java:
```diff
@@ -0,0 +1,510 @@
+     * @return a {@link RetainableByteBuffer} containing the resource's contents.
+     * @throws IllegalArgumentException if the resource is a directory or does not exist or there is no way to access its contents.
+     */
+    public static RetainableByteBuffer toRetainableByteBuffer(Resource resource, ByteBufferPool bufferPool, boolean direct) throws IllegalArgumentException
```
Note that with my RBB changes, then this method could delay any reading of the actual resource until need either in `toByteByffer` or `writeTo(Content.Sink)`.  In the later case, the whole resource would never need to be in memory at once.

just saying....

lorban replied:
The only non-test usage of this method is in `CachingHttpContentFactory` where we want to keep an in-memory copy of the resource for caching purposes.

I'm not sure how the RBB changes would help here?

gregw replied:
I'm just saying that if there was a need, then we could hide a whole resource reader within the RBB abstraction, and so long as nobody called getByteBuffer, then we'd never have the whole resource in memory.     I.e. the power of using RBB as the primary interface rather than RBB.getByteBuffer

gregw commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/IOResources.java:
```diff
@@ -0,0 +1,510 @@
+        }
+
+        // Fallback to InputStream.
+        try (InputStream inputStream = resource.newInputStream())
```
Is this code testable?  Do we have any non-directory resources that do not have a Path?  Maybe ISE until we know it is needed...  or perhaps test with a URLResource?

lorban replied:
It is testable with `URLResourceFactory`. Let me modify the tests.

lorban committed:
add tests and fix bugs

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

gregw commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/Resource.java:
```diff
@@ -238,13 +238,12 @@ public InputStream newInputStream() throws IOException
      * @throws IOException if unable to open the readable bytechannel for the resource.
+     * @deprecated use {@link #newInputStream()} or {@code IOResources} instead.
      */
+    @Deprecated(since = "12.0.8", forRemoval = true)
```
should we deprecate `newInputStream` as well?

lorban replied:
If we do, this means two things:

 - `IOResources` does not have any fallback in case it doesn't know about the resource type.
 - `jetty-util` and `jetty-xml` uses of that method have to go away. Tests can probably safely assume that a resource has a path, but `BufferUtil.toBuffer(Resource, boolean)`, `CertificateUtils`, `XmlConfiguration` and other utilities would need the same kind of logic that `IOResources` has.

gregw replied:
OK. leave it as is for now.

gregw commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/Resource.java:
```diff
@@ -238,13 +238,12 @@ public InputStream newInputStream() throws IOException
-        if (path == null)
-            return null;
-        return Files.newByteChannel(getPath(), StandardOpenOption.READ);
+       return null;
```
let's leave the implementation for now.

lorban replied:
OK.

gregw commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/MemoryResource.java:
```diff
@@ -37,7 +35,7 @@ public class MemoryResource extends Resource
     private final URI _uri;
     private final Instant _created = Instant.now();
-    private final byte[] _bytes;
+    private final byte[] _bytes; // TODO a ByteBuffer would be nicer
```
Can that be done?

lorban replied:
That's actually not needed. I've removed that comment.

lorban committed:
add null check to use non-pooling bufferPool, improve log, remove outdated todos

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
handle review comments

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
fix test

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/IOResources.java:
```diff
@@ -0,0 +1,516 @@
+
+    private static class PathToSinkCopier extends IteratingNestedCallback
+    {
+        private static final Logger LOG = LoggerFactory.getLogger(PathToSinkCopier.class);
```
Never used, remove.

sbordet reviewed (dismissed):
Just one nit, LGTM.

lorban committed:
remove unneeded logger

Signed-off-by: Ludovic Orban <lorban@bitronix.be>
